var/spack/repos/builtin/packages/intel-oneapi-mpi/package.py

```diff
@@ -50,3 +54,47 @@ def libs(self):
             ldir = find_libraries('*', root=lib_path, shared=True, recursive=False)
             libs += ldir
         return libs
+
+    def _join_prefix(self, path):
+        return join_path(self.prefix, 'mpi', 'latest', path)
+
+    def _ld_library_path(self):
+        dirs = ['lib',
+                'lib/release',
+                'libfabric/lib']
+        for dir in dirs:
+            yield self._join_prefix(dir)
+
+    def _library_path(self):
+        dirs = ['lib',
+                'lib/release',
+                'libfabric/lib']
+        for dir in dirs:
+            yield self._join_prefix(dir)
+
+    def install(self, spec, prefix):
+        super(IntelOneapiMpi, self).install(spec, prefix)
+
+        # need to patch libmpi.so so it can always find libfabric
+        libfabric_rpath = self._join_prefix('libfabric/lib')
+        for lib_version in ['debug', 'release', 'release_mt', 'debug_mt']:
+            file = self._join_prefix('lib/' + lib_version + '/libmpi.so')
+            subprocess.call(['patchelf', '--set-rpath', libfabric_rpath, file])
+
+    def setup_run_environment(self, env):
+        env.prepend_path('PATH', self._join_prefix('bin'))
+        env.prepend_path('CPATH', self._join_prefix('include'))
+        for dir in self._library_path():
+            env.prepend_path('LIBRARY_PATH', dir)
+        for dir in self._ld_library_path():
+            env.prepend_path('LD_LIBRARY_PATH', dir)
+        # set these so that wrappers know what compilers to use
+        env.set('I_MPI_CC', 'icx')
+        env.set('I_MPI_CXX', 'icpx')
+        env.set('I_MPI_FC', 'ifx')
```

Comment:
According to the help message that the wrapper prints, e.g. for mpicc:
```
 24     echo "   1. Command line option:  -cc=<compiler_name>"
 25     echo "   2. Environment variable: I_MPI_CC (current value '$I_MPI_CC')"
 26     echo "   3. Environment variable: MPICH_CC (current value '$MPICH_CC')"
```
either will be accepted, with MPICH_CC having the lowest precedent. I don't have a specific requirement for one over the other, just that one of these sets of variables needs to be set in the module. 

Having them specified here, rather than in the dependent_build_environment, ensures that these wrappers will work correctly when invoked from outside of spack, and this is a use case we care about. As much as we like spack for building, sometimes we just want the compiler.

Reply:
Thanks for the information! This all looks good then, although I'm curious why you didn't have to set `MPICC` etc.

Reply:
Thanks for looking over it. MPICC isn't set when sourcing setvars.sh either. What specifically consumes MPICC? CMake findMPI or something similar?

Reply:
I don't see it as a variable in https://cmake.org/cmake/help/latest/module/FindMPI.html, but I thought generally `MPICC` is how dependent packages could find the `mpicc` wrapper (and generally `MPICH_CC`/`I_MPI_CC` was how the `mpicc` wrapper found the underlying compiler). Perhaps CMake can locate the `mpicc` wrapper with `I_MPI_root`, but the Spack `openmpi` package for example sets `MPICC` in the environment.

Reply:
I don't understand why we would set I_MPI_CC in this package. Shouldn't the compiler package set the variables? What if you want to use intel mpi wrappers and gcc as the underlying compiler?

Reply:
Good point. I see two solutions: 

- Set the variable conditionally when spec contains '%oneapi'
- Set the variable with the compiler

I'm leaning toward setting with the compiler. For us, this will eventually form part of a Cray PrgEnv, and having the oneapi compiler loaded would be a good reason to set this value.

Reply:
Not knowing spack, I am still confused. In an earlier commit I added setup_dependent_build_environment for the MPI package to set MPICH_CC to spack_cc. I thought this would make the wrapper use the compiler that was part of the spec. Is this accurate? Is a case not covered?

Reply:
> I added setup_dependent_build_environment for the MPI package to set MPICH_CC to spack_cc.

`setup_run_environment` runs under different conditions: unlike `setup_dependent_build_environment` is used to generate the module file and to set environment variables when you run `spack env activate ...`. So regarding 

> I thought this would make the wrapper use the compiler that was part of the spec. Is this accurate?

Yes

This is specifically to address https://github.com/spack/spack/pull/20808/files#r562032811

> [we want to ensure] that these wrappers will work correctly when invoked from outside of spack

Reply:
I moved the I_MPI_{CC,CXX,FC} environment settings to the intel-oneapi-compilers module as this seems to make the most sense for our expected usage patterns.